loopout.go

```diff
@@ -706,7 +706,77 @@ func (s *loopOutSwap) waitForHtlcSpendConfirmed(globalCtx context.Context,
 		return nil, fmt.Errorf("register spend ntfn: %v", err)
 	}
 
+	var paymentComplete bool
+	pushPreimage := func(ctx context.Context) error {
+		// If our payment is complete, there is no need to proceed in
+		// attempting to push the preimage.
+		if paymentComplete {
+			return nil
+		}
+
+		// If we have not yet revealed our preimage through a sweep, we
+		// do not push the preimage because we may choose to never sweep
+		// if fees are too high.
+		if s.state != loopdb.StatePreimageRevealed {
```

Comment:
By the time we get here, haven't we always revealed already?

Reply:
We might have called our sweep function but not actually decided to sweep because fees are too high.

Reply:
Ah, that is also what the comment says indeed.

And what if we persist the revealed state and crash before we actually publish (reveal)? On restart we are going to disclose the unpublished preimage to the server?

Reply:
I don't think it matters because our current sweep logic will take note of the reveal and publish at all costs? 
If we restart with preimage revealed state, the sweeper will always publish our sweep with the max fee. 

There is the case where the sweep still doesn't get in (maybe high fees, maybe bitcoind borked) but I think that should be addressed by more aggressive client sweeping? We also won't sweep after preimage push (not that that should be relied on).

Reply:
Suppose we have no more backend after the crash?

Reply:
Then the off chain htlc will be pulled and the on chain will not be claimed. 
But I think this falls into the category of "exceptional errors" that we can't really account for. In that case the swapper would need to manually sweep their output. 

We have no way to guarantee publish/mempool propagation other than to just keep trying to sweep, which we already do.

Reply:
Yes strong point. Publish isn't a guarantee either indeed.

Reply:
Then the off chain htlc will be pulled and the on chain will not be claimed. 
But I think this falls into the category of "exceptional errors" that we can't really account for. In that case the swapper would need to manually sweep their output. 

We have no way to guarantee publish/mempool propagation other than to just keep trying to sweep, which we already do.

Reply:
Thinking about this again, isn't this inconsistent? So the happy flow is:
1. Mark revealed in db
2. Reveal in sweep
3. Reveal to server endpoint

However if we crash right after 1 and restart, the order becomes:
1. Mark revealed in db
2. Reveal to server endpoint
3. Reveal in sweep

To minimize variation, isn't it better to always stick to the second sequence?

One other thing we haven't discussed is failure to publish the tx because of a bug in loop and/or wallet code. In that case, the first sequence is safer. To always do the first sequence, I think we'd need two states: `PreimageRevealPublish` and  `PreimageRevealed`. Can also be useful in certain debug scenarios perhaps.

Reply:
How would the second sequence you list ever happen? If we mark preimage as revealed and then crash, we'll call `sweep` with our state set to `PreimageRevealed` on restart. We will broadcast a sweep tx again before we push preimage to server?

Reply:
I was assuming the push preimage closure was ran even before the main loop, but ofc that is just the definition of the closure. Resolved

Reply:
Moving this out of a closure into its own function so it reads a bit better as well.